src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+    const errorPrefix = isPrimaryAddress ? this.primaryAddressPrefix : this.secondaryAddressPrefix;
+
+    return {
+      postcodeValid: postcodeErrors.length === 0,
+      addressValid: addressErrors.length === 0,
+      errors: typeErrors.concat(addressErrors).concat(postcodeErrors).map(error => errorPrefix + error)
+    };
+  }
+
+  private validateType(address: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    if (isPrimaryAddress || (!!address.postcode || this.addressFieldsNotEmpty(address))) {
+      return !address.type_id ? [this.typeRequiredError] : [];
+    }
+    return [];
+  }
+
+  private validateAddressLines(address: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    const validateSecondaryAddress = (!isPrimaryAddress && (this.addressFieldsNotEmpty(address) || address.postcode?.trim()));
+    const errors: string[] = [];
+
+    if (isPrimaryAddress || validateSecondaryAddress) {
+      if (!address.address_lines?.trim()) {
+        errors.push(this.addressRequiredError);
+      }
+      if(!address.town?.trim()) {
+        errors.push(this.townRequiredError);
+      }
+    }
+    return errors;
+  }
+
+  private validatePostcode(courtAddress: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    const errors: string[] = [];
+    const postcodeRequired = isPrimaryAddress || (!isPrimaryAddress && this.addressFieldsNotEmpty(courtAddress));
+    const postcodeEmpty = !courtAddress?.postcode?.trim();
+
+    if (postcodeRequired && postcodeEmpty) {
+      errors.push(this.postcodeMissingError);
+    }
+
+    if (!postcodeEmpty && !this.postcodeIsValidFormat(courtAddress.postcode)) {
+      errors.push(this.invalidPostcodeError);
+    }
+
+    return errors;
+  }
+
+  private validateAddressTypes(primary: DisplayAddress, secondary: DisplayAddress, writeToUsTypeId: number): string[] {
+    return writeToUsTypeId && !!primary.type_id && !!secondary.type_id && // validate only if types selected in both addresses
+    (this.addressFieldsNotEmpty(secondary) || secondary.postcode?.trim()) && // validate only if secondary has some fields entered
+    (primary.type_id !== writeToUsTypeId && secondary.type_id !== writeToUsTypeId) // at least 1 write address should exist
+      ? [this.multipleVisitAddressError]
+      : [];
+  }
+
+  private addressFieldsNotEmpty(courtAddress: DisplayAddress): boolean {
+    return !!courtAddress.address_lines?.trim() || !!courtAddress.address_lines_cy?.trim() ||
+      !!courtAddress.town?.trim() || !!courtAddress.town_cy?.trim();
+  }
+
+  private postcodeIsValidFormat(postcode: string): boolean {
+    const postcodeRegex = new RegExp('([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|' +
+      '(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))' +
+      '\\s?[0-9][A-Za-z]{2})', 'g');
+
+    const match = postcode.match(postcodeRegex);
+    return match?.length === 1 && match[0] === postcode;
```

Comment:
Replacing the two lines with:

`return postcode.match(postcodeRegex) ? match[0] === postcode : false` might be simpler (I think match returns a boolean?)? Can prob be shortened further if you want

Reply:
String.match() returns an array of matches (or null if no match found) so we'd be unable to use a ternary operator here. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match